src/variety/tents.js

```diff
@@ -0,0 +1,499 @@
+				if (cell.getNum()!==2) {
+					continue;
+				}
+				var target = null,
+					clist = new this.klass.CellList();
+				// 右・左下・下・右下だけチェック
+				clist.add(cell);
+				target = cell.relcell(2, 0);
+				if (target.getNum()===2) {
+					clist.add(target);
+				}
+				target = cell.relcell(0, 2);
+				if (target.getNum()===2) {
+					clist.add(target);
+				}
+				target = cell.relcell(-2, 2);
+				if (target.getNum()===2) {
+					clist.add(target);
+				}
+				target = cell.relcell(2, 2);
+				if (target.getNum()===2) {
+					clist.add(target);
+				}
+				if (clist.length <= 1) {
+					continue;
+				}
+
+				this.failcode.add("tentAround");
+				if (this.checkOnly) {
+					break;
+				}
+				clist.seterr(1);
+			}
+		},
+
+		checkTentNone: function() {
+			this.checkCamps_inner(function(area) {
+				return area.counts.tents === 0;
+			}, "nmTentNone");
+		},
+
+		checkTreeNone: function() {
+			this.checkCamps_inner(function(area) {
+				return area.counts.trees === 0;
+			}, "nmTreeNone");
+		},
+
+		checkCampsLt: function() {
+			this.checkCamps_inner(function(area) {
+				return area.counts.tents > 0 && area.counts.tents < area.counts.trees;
+			}, "nmTentLt");
+		},
+
+		checkCampsGt: function() {
+			this.checkCamps_inner(function(area) {
+				return area.counts.trees > 0 && area.counts.tents > area.counts.trees;
+			}, "nmTentGt");
+		},
+
+		checkCamps_inner: function(func, code) {
+			var rooms = this.board.campgraph.components;
+			for (var id = 0; id < rooms.length; id++) {
+				var area = rooms[id];
+				if (!area || !func(area)) { continue; }
+
+				this.failcode.add(code);
+				if (this.checkOnly) {
+					break;
+				}
+				area.clist.seterr(1);
+			}
+		},
+
+		checkTentCount: function() {
+			this.checkRowsCols(this.isExCellCount, "exTentNe");
+		},
+
+		isExCellCount: function(clist) {
+			var d = clist.getRectSize(),
+				bd = this.board;
+			var count = clist.filter(function(c) {
+				return c.getNum()===2;
+			}).length;
+
+			var result = true;
+
+			if (d.x1 === d.x2) {
+				var exc = bd.getex(d.x1, -1);
+				if (exc.qnum !== -1 && exc.qnum !== count) {
+					exc.seterr(1);
+					result = false;
+				}
+			}
+			if (d.y1 === d.y2) {
+				var exc = bd.getex(-1, d.y1);
+				if (exc.qnum !== -1 && exc.qnum !== count) {
+					exc.seterr(1);
+					result = false;
+				}
+			}
+
+			if (!result) {
+				clist.seterr(1);
+			}
+			return result;
+		}
+	},
+
+	FailCode: {
+		nmTentNone: [
+			"(please translate) A tree has no tent.",
+			"A tree has no tent."
+		],
+		nmTreeNone: [
+			"(please translate) A tent is not next to a tree.",
+			"A tent is not next to a tree."
+		],
+		nmTentLt: [
+			"(please translate) There aren't enough tents around a tree.",
+			"There aren't enough tents around a tree."
+		],
+		nmTentGt: [
+			"(please translate) There are too many tents around a tree.",
+			"There are too many tents around a tree."
+		],
+		tentAround: [
+			"(please translate) Tents are adjacent.",
+			"Tents are adjacent."
```

Comment:
I think (some) people read "adjacent" as touch by an edge. Would suggest "Some tents touch.", as we have with star battle.